openstack/nova/shellinabox/revproxy.lua

```diff
@@ -0,0 +1,226 @@
+
+    local errors = {}
+
+    for i = 1,n_threads do
+        local ok, res = ngx.thread.wait(threads[i])
+        if not ok then
+            table.insert(errors, "Could not query thread due to " .. res)
+        elseif res then
+            if res[1] then return res[1] end
+            if res[2] then table.insert(errors, err) end
+        end
+    end
+
+    -- We may be without result due to an error (or more)
+    if #errors > 0 then return nil, "Lookup failed due to: " .. table.concat(errors, "\n")  end
+    ngx.log(ngx.INFO, "No results for hash " .. token_hash)
+end
+
+function _M.lookup_memc(token_hash)
+    local memc, err = memcached:new()
+    if not memc then return nil, "failed to instantiate memc: " .. err end
+    memc:set_timeout(1000) -- 1 sec
+
+    local ok, err = memc:connect(config.memcached[1], config.memcached[2])
+    if not ok then return nil, "failed to connect: " .. err end
+
+    local res, flags, err = memc:get(token_hash)
+    if err then ngx.log(ngx.WARN, "failed to get token from memcached due to: ", err) end
+    if res then
+        logged_keepalive(memc)
+        return cjson.decode(res)
+    end
+
+    res = _M.lookup_db(token_hash)
+    memc:set(token_hash, cjson.encode(res), 10)
+    logged_keepalive(memc)
+    return res
+end
+
+_M.lookup = store_in_ctx("entry", function()
+    local token_hash = sha256_hex(_M.get_token())
+    if not token_hash then
+        ngx.log(ngx.ERR, "Could not hash token")
+        return ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
+    end
+
+    local entry, err
+    if not _M.cache then
+        entry, err = _M.lookup_memc(token_hash)
+    else
+        entry, err = _M.cache:get(token_hash, nil, _M.lookup_memc, token_hash)
+    end
+
+    if err then
+        ngx.log(ngx.ERR, "could not retrieve user: ", err)
+        return ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
+    end
+    return entry
+end)
+
+function _M.handle_redirect()
+    local location = ngx.header.Location
+    if not location then return end
+
+    local token = _M.get_token()
+    if not token then return end
+    local token_base = url_base .. token
+
+    local self_host = ngx.var.host
+    local scheme, host, path, query = location:match("([^:]+)://([^/]+)([^?]+)(?%g+)")
+    local new_q = {}
+    for key, value in query:sub(2):gmatch("([^=&]+)=([^&]*)") do
+        if key == "ip" then value = self_host
+        elseif key == "kvmport" then value = "443" .. token_base
+        elseif key == "title" then value = ngx.ctx.entry.instance_uuid or "KVM"
+        end
+        table.insert(new_q, key .. "=" .. value)
+    end
+    local new_query = table.concat(new_q, "&")
+
+    local new_location = token_base .. path .. "?" .. new_query
+    ngx.header.Location = new_location
+end
+
+function _M.process()
+    local entry = _M.lookup()
+    if not entry then return end
+    local host = host_from_entry(entry)
+    local request_uri = ngx.var.request_uri
+    local sub_uri = request_uri:sub(38)
```

Comment:
Can you point me to any docs for this `sub()`? I can only find `string.sub()` and would like to find out what the `38` might mean here.

Reply:
I've added some more comments, I hope that clarifies things a bit more